File: Task2/Task2_Code/Container.cpp
//
// Created by Vsevolod Melnik on 15.10.2021.
//

#include "Container.h"

Container *Container::In(FILE *file) {
    Container* c = new Container();
    int k;
    fscanf(file, "%d", &k);
    if (k < c->maxLength) {
        c->length = k;
        for (int i = 0; i < c->length; ++i)
            c->shapes[i] = Shape::In(file);
    } else {
        printf("Reading error: too many shapes.");
    }
    return c;
}

void Container::Out(FILE *file) {
    fprintf(file, "%d shapes:", length);
    printf("%d shapes:", length);
    for (int i = 0; i < length; ++i) {
        fprintf(file, "\n");
        printf("\n");
        shapes[i]->Out(file);
    }
}

Container *Container::InitRandom(int cnt) {
    Container* c = new Container();
    if (cnt > Container::maxLength) {
        printf("Too much shapes were requested. The program will generate only %d\n", Container::maxLength);
        cnt = Container::maxLength;
    }
    int numberOfShapes = cnt;
    c->length = numberOfShapes;
    for (int i = 0; i < numberOfShapes; ++i) {
        c->shapes[i] = Shape::InitRandom();
    }
    return c;
}

void Container::DeleteLessThenAverage() {
    double averageArea = 0;
    for (int i = 0; i < length; ++i)
        averageArea += shapes[i]->FindArea();
    if (length <= 0)
        return;
    averageArea /= length;
    int lastToStay = 0;
    for (int i = 0; i < length; ++i) {
        if (shapes[i]->FindArea() >= averageArea)
            shapes[lastToStay++] = shapes[i];
    }
    length = lastToStay;
}

File: Task3/Task3_Code/container.py
from shape import Shape


# Container of shapes.
class Container:

    max_length = 10000

    def __init__(self):
        self.shapes = []
        self.number_of_shapes = 0

    @staticmethod
    def init_random(cnt):
        if cnt > Container.max_length:
            print("Too much shapes were requested. The program will generate only", Container.max_length)
            cnt = Container.max_length
        container = Container()
        container.number_of_shapes = cnt
        for i in range(cnt):
            container.shapes.append(Shape.init_random())
        return container

    def output(self, file_name):
        with open(file_name, "w") as f:
            print("Number of shapes: " + str(self.number_of_shapes))
            f.write("Number of shapes: " + str(self.number_of_shapes) + "\n")
            for i in range(len(self.shapes)):
                self.shapes[i].output(f)

    @staticmethod
    def input(file_name):
        with open(file_name, "r") as f:
            cnt = f.readline()[:-1]
            if (not cnt.isdigit()) or (int(cnt) < 0 or int(cnt) > Container.max_length):
                print("Reading error: expected number of shape.")
                return
            container = Container()
            container.number_of_shapes = int(cnt)
            for i in range(container.number_of_shapes):
                args = map(int, f.readline().split())
                container.shapes.append(Shape.input(args))
            return container

    # Delete shapes, whose area is less then average.
    def delete_less_then_average(self):
        if self.number_of_shapes <= 0:
            return
        average_area = sum(s.find_area() for s in self.shapes) / self.number_of_shapes
        new_shapes = []
        for i in range(self.number_of_shapes):
            if self.shapes[i].find_area() >= average_area:
                new_shapes.append(self.shapes[i])
        self.shapes = new_shapes
        self.number_of_shapes = len(self.shapes)
